src/java.base/share/classes/java/lang/reflect/code/parser/impl/DescParser.java

```diff
@@ -85,16 +88,42 @@ public static RecordTypeRef parseRecordTypeRef(String desc) {
     }
 
     public static TypeDefinition parseTypeDefinition(Lexer l) {
+        boolean isTypeVar = false;
+        if (l.token().kind == TokenKind.HASH) {
+            l.accept(TokenKind.HASH);
+            isTypeVar = true;
+        }
         // Type
-        Tokens.Token t = l.accept(Tokens.TokenKind.IDENTIFIER);
+        Tokens.Token t = l.accept(TokenKind.IDENTIFIER,
+                TokenKind.PLUS, TokenKind.SUB);
         StringBuilder identifier = new StringBuilder();
-        identifier.append(t.name());
+        identifier.append(t.kind == TokenKind.IDENTIFIER ? t.name() : t.kind.name);
         while (l.acceptIf(Tokens.TokenKind.DOT)) {
             identifier.append(Tokens.TokenKind.DOT.name);
             t = l.accept(Tokens.TokenKind.IDENTIFIER);
             identifier.append(t.name());
         }
 
+        if (l.token().kind == TokenKind.COLCOL && isTypeVar) {
+            // type-variable
+            l.accept(TokenKind.COLCOL);
+            identifier.append(TokenKind.COLCOL.name);
+            t = l.accept(TokenKind.IDENTIFIER); // type-var or method name
+            identifier.append(t.name());
+            if (l.token().kind == TokenKind.LPAREN) {
+                FunctionType functionType = parseMethodType(l);
```

Comment:
Note that here we parse, then we throw away, as the type definition only wants a string-based identifier, so we'll need to reparse the identifier string again in the type factory.